From dd2ffd141b0df5eea092b3e7189f450af43571bc Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Thu, 13 Sep 2018 22:58:21 +0300
Subject: [PATCH] LVDS: update AR0233, AR0143, GW4200

1) update AR0233 sensor with OTP/resolution/controls
2) clean-up AR0143/GW4200
3) change AR0143 HTS/VTS to set 4 camera setup stable
---
 drivers/media/i2c/soc_camera/ar0143.c        |  8 ++---
 drivers/media/i2c/soc_camera/ar0143.h        |  6 ++--
 drivers/media/i2c/soc_camera/ar0233.c        | 52 +++++++++++++++++++---------
 drivers/media/i2c/soc_camera/ar0233.h        | 47 +++++++++++++++----------
 drivers/media/i2c/soc_camera/gw4200_ar014x.c |  2 ++
 5 files changed, 72 insertions(+), 43 deletions(-)

diff --git a/drivers/media/i2c/soc_camera/ar0143.c b/drivers/media/i2c/soc_camera/ar0143.c
index 1a6ec5b..5da1271 100644
--- a/drivers/media/i2c/soc_camera/ar0143.c
+++ b/drivers/media/i2c/soc_camera/ar0143.c
@@ -23,7 +23,6 @@
 #include "ar0143.h"
 
 #define AR0143_I2C_ADDR		0x10
-//#define AR0143_I2C_ADDR		0x54 // eeprom
 
 #define AR0143_PID		0x3000
 #define AR0143_VERSION_REG	0x0D54
@@ -37,7 +36,6 @@ struct ar0143_priv {
 	struct v4l2_rect		rect;
 	int				init_complete;
 	u8				id[6];
-	int				dvp_order;
 	/* serializers */
 	int				max9286_addr;
 	int				max9271_addr;
@@ -145,7 +143,6 @@ static int ar0143_set_fmt(struct v4l2_subdev *sd,
 {
 	struct v4l2_mbus_framefmt *mf = &format->format;
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
-	struct ar0143_priv *priv = to_ar0143(client);
 
 	mf->code = AR0143_MEDIA_BUS_FMT;
 	mf->colorspace = V4L2_COLORSPACE_SMPTE170M;
@@ -338,6 +335,9 @@ static int ar0143_s_ctrl(struct v4l2_ctrl *ctrl)
 			val &= ~0x8000;
 		ret |= reg16_write16(client, 0x3040, val);
 		break;
+	case V4L2_CID_MIN_BUFFERS_FOR_CAPTURE:
+		ret = 0;
+		break;
 	}
 
 	return ret;
@@ -469,8 +469,6 @@ static int ar0143_parse_dt(struct device_node *np, struct ar0143_priv *priv)
 
 		of_node_put(endpoint);
 
-		of_property_read_u32(endpoint, "dvp-order", &priv->dvp_order);
-
 		rendpoint = of_parse_phandle(endpoint, "remote-endpoint", 0);
 		if (!rendpoint)
 			continue;
diff --git a/drivers/media/i2c/soc_camera/ar0143.h b/drivers/media/i2c/soc_camera/ar0143.h
index cc5172a..c6b1014 100644
--- a/drivers/media/i2c/soc_camera/ar0143.h
+++ b/drivers/media/i2c/soc_camera/ar0143.h
@@ -491,8 +491,10 @@ struct ar0143_reg {
 {0x3C08, 0x0100}, // RESERVED_MFR_3C08
 {0x3C0C, 0x0518}, // DELAY_BUFFER_LLPCK_RD_WR_OVERLAP
 #ifdef NEW_TBL
-{0x300A, 0x0438}, // FRAME_LENGTH_LINES_
-{0x300C, 0x060e}, // LINE_LENGTH_PCK_
+//{0x300A, 0x0438}, // FRAME_LENGTH_LINES_
+//{0x300C, 0x060e}, // LINE_LENGTH_PCK_
+{0x300A, AR0143_SENSOR_HEIGHT + 157}, // FRAME_LENGTH_LINES_
+{0x300C, AR0143_SENSOR_WIDTH + 144}, // LINE_LENGTH_PCK_
 #else
 {0x300A, 0x048A}, // FRAME_LENGTH_LINES_
 {0x300C, 0x05BA}, // LINE_LENGTH_PCK_
diff --git a/drivers/media/i2c/soc_camera/ar0233.c b/drivers/media/i2c/soc_camera/ar0233.c
index 0ac8f3a..b392ff9 100644
--- a/drivers/media/i2c/soc_camera/ar0233.c
+++ b/drivers/media/i2c/soc_camera/ar0233.c
@@ -23,13 +23,11 @@
 #include "ar0233.h"
 
 #define AR0233_I2C_ADDR		0x10
-//#define AR0233_I2C_ADDR		0x54 // eeprom
 
 #define AR0233_PID		0x3000
 #define AR0233_VERSION_REG	0x0354
 
-//#define AR0233_MEDIA_BUS_FMT	MEDIA_BUS_FMT_SGRBG12_1X12
-#define AR0233_MEDIA_BUS_FMT	MEDIA_BUS_FMT_SBGGR12_1X12
+#define AR0233_MEDIA_BUS_FMT	MEDIA_BUS_FMT_SGRBG12_1X12
 
 struct ar0233_priv {
 	struct v4l2_subdev		sd;
@@ -38,9 +36,6 @@ struct ar0233_priv {
 	struct v4l2_rect		rect;
 	int				init_complete;
 	u8				id[6];
-	int				exposure;
-	int				gain;
-	int				autogain;
 	/* serializers */
 	int				ti9x4_addr;
 	int				ti9x3_addr;
@@ -262,18 +257,26 @@ static int ar0233_s_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_GAMMA:
 	case V4L2_CID_SHARPNESS:
 	case V4L2_CID_AUTOGAIN:
+		break;
 	case V4L2_CID_GAIN:
+		/* Digital gain */
+		ret = reg16_write16(client, 0x3308, ctrl->val);
+		break;
+	case V4L2_CID_ANALOGUE_GAIN:
+		/* Analog gain */
+		ret = reg16_write16(client, 0x3366, (ctrl->val << 8) | (ctrl->val << 4) | ctrl->val);
+		break;
 	case V4L2_CID_EXPOSURE:
+		/* T1 exposure */
+		ret = reg16_write16(client, 0x3012, ctrl->val);
 		break;
 	case V4L2_CID_HFLIP:
 		ret = reg16_read16(client, 0x3040, &val);
-		if (ret < 0)
-			goto out;
 		if (ctrl->val)
 			val |= (1 << 14);
 		else
 			val &= ~(1 << 14);
-		ret = reg16_write16(client, 0x3040, val);
+		ret |= reg16_write16(client, 0x3040, val);
 		break;
 	case V4L2_CID_VFLIP:
 		ret = reg16_read16(client, 0x3040, &val);
@@ -281,14 +284,13 @@ static int ar0233_s_ctrl(struct v4l2_ctrl *ctrl)
 			val |= (1 << 15);
 		else
 			val &= ~(1 << 15);
-		ret = reg16_write16(client, 0x3040, val);
+		ret |= reg16_write16(client, 0x3040, val);
 		break;
 	case V4L2_CID_MIN_BUFFERS_FOR_CAPTURE:
 		ret = 0;
 		break;
 	}
 
-out:
 	return ret;
 }
 
@@ -318,6 +320,21 @@ static int ar0233_s_ctrl(struct v4l2_ctrl *ctrl)
 
 static void ar0233_otp_id_read(struct i2c_client *client)
 {
+	struct ar0233_priv *priv = to_ar0233(client);
+	int i;
+	u16 val = 0;
+
+	/* read camera id from ar014x OTP memory */
+	reg16_write16(client, 0x3054, 0x400);
+	reg16_write16(client, 0x304a, 0x110);
+	usleep_range(25000, 25500); /* wait 25 ms */
+
+	for (i = 0; i < 6; i += 2) {
+		/* first 4 bytes are equal on all ar014x */
+		reg16_read16(client, 0x3800 + i + 4, &val);
+		priv->id[i]     = val >> 8;
+		priv->id[i + 1] = val & 0xff;
+	}
 }
 
 static ssize_t ar0233_otp_id_show(struct device *dev,
@@ -327,6 +344,8 @@ static ssize_t ar0233_otp_id_show(struct device *dev,
 	struct i2c_client *client = v4l2_get_subdevdata(sd);
 	struct ar0233_priv *priv = to_ar0233(client);
 
+	ar0233_otp_id_read(client);
+
 	return snprintf(buf, 32, "%02x:%02x:%02x:%02x:%02x:%02x\n",
 			priv->id[0], priv->id[1], priv->id[2], priv->id[3], priv->id[4], priv->id[5]);
 }
@@ -439,9 +458,6 @@ static int ar0233_probe(struct i2c_client *client,
 	v4l2_i2c_subdev_init(&priv->sd, client, &ar0233_subdev_ops);
 	priv->sd.flags = V4L2_SUBDEV_FL_HAS_DEVNODE;
 
-	priv->exposure = 0x100;
-	priv->gain = 0x100;
-	priv->autogain = 1;
 	v4l2_ctrl_handler_init(&priv->hdl, 4);
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
 			  V4L2_CID_BRIGHTNESS, 0, 16, 1, 7);
@@ -456,11 +472,13 @@ static int ar0233_probe(struct i2c_client *client,
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
 			  V4L2_CID_SHARPNESS, 0, 10, 1, 3);
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
-			  V4L2_CID_AUTOGAIN, 0, 1, 1, priv->autogain);
+			  V4L2_CID_AUTOGAIN, 0, 1, 1, 0);
+	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
+			  V4L2_CID_GAIN, 1, 0x7ff, 1, 0x200);
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
-			  V4L2_CID_GAIN, 0, 0xffff, 1, priv->gain);
+			  V4L2_CID_ANALOGUE_GAIN, 1, 0xe, 1, 0xa);
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
-			  V4L2_CID_EXPOSURE, 0, 0xffff, 1, priv->exposure);
+			  V4L2_CID_EXPOSURE, 1, 0x600, 1, 0x144);
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
 			  V4L2_CID_HFLIP, 0, 1, 1, 0);
 	v4l2_ctrl_new_std(&priv->hdl, &ar0233_ctrl_ops,
diff --git a/drivers/media/i2c/soc_camera/ar0233.h b/drivers/media/i2c/soc_camera/ar0233.h
index bd67ece..933d3d1 100644
--- a/drivers/media/i2c/soc_camera/ar0233.h
+++ b/drivers/media/i2c/soc_camera/ar0233.h
@@ -1,5 +1,5 @@
 /*
- * ON Semiconductor AR0233 sensor camera wizard 1920x1080@44/RCCB/MIPI
+ * ON Semiconductor AR0233 sensor camera wizard 1920x1080@30/BGGR/MIPI
  *
  * Copyright (C) 2018 Cogent Embedded, Inc.
  *
@@ -12,11 +12,19 @@
 //#define AR0233_DISPLAY_PATTERN_FIXED
 //#define AR0233_DISPLAY_PATTERN_COLOR_BAR
 
-#define AR0233_MAX_WIDTH	1808
-#define AR0233_MAX_HEIGHT	1150
+#define AR0233_MAX_WIDTH	1792 //1920
+#define AR0233_MAX_HEIGHT	1080
 
 #define AR0233_DELAY		0xffff
 
+#define AR0233_SENSOR_WIDTH	1920
+#define AR0233_SENSOR_HEIGHT	1280
+
+#define AR0233_X_START		((AR0233_SENSOR_WIDTH - AR0233_MAX_WIDTH) / 2)
+#define AR0233_Y_START		((AR0233_SENSOR_HEIGHT - AR0233_MAX_HEIGHT) / 2)
+#define AR0233_X_END		(AR0233_X_START + AR0233_MAX_WIDTH - 1)
+#define AR0233_Y_END		(AR0233_Y_START + AR0233_MAX_HEIGHT + 1) /* must be +1 and not -1 or 2 lines missed - bug in imager? */
+
 struct ar0233_reg {
 	u16	reg;
 	u16	val;
@@ -335,26 +343,25 @@ struct ar0233_reg {
 {0x33E4, 0x0080},
 {0x33E0, 0x0C80},
 {0x33E0, 0x0C80},
-{0x3004, 0x002E},
-{0x3008, 0x073D},
-{0x3002, 0x0019},
-{0x3006, 0x0496},
+{0x3004, AR0233_X_START}, // X_ADDR_START_
+{0x3008, AR0233_X_END}, // X_ADDR_END_
+{0x3002, AR0233_Y_START}, // Y_ADDR_START_
+{0x3006, AR0233_Y_END}, // Y_ADDR_END_
+{0x3402, 0x0000 | AR0233_MAX_WIDTH}, // X_OUTPUT_CONTROL
+{0x3404, 0x0000 | AR0233_MAX_HEIGHT}, // Y_OUTPUT_CONTROL
 {0x3032, 0x0000},
 {0x3400, 0x0010},
-{0x3402, 0x0780},
-{0x3404, 0x04B8},
+#if 1
+/* disable HDR */
 {0x3082, 0x0000},
-{0x30BA, 0x11F1},
-
+{0x30BA, 0x11F0},
+#endif
 {AR0233_DELAY, 100}, // Wait 100ms
 
-{0x30BA, 0x11F0},
-{0x300C, 0x043E},
-{0x300A, 0x0A8C},
+{0x300A, AR0233_SENSOR_HEIGHT + 356}, // FRAME_LENGTH_LINES_
+{0x300C, AR0233_SENSOR_WIDTH + 100}, // LINE_LENGTH_PCK_
 {0x3042, 0x0000},
 {0x3238, 0x0222},
-{0x3238, 0x0222},
-{0x3238, 0x0222},
 {0x3012, 0x0144},
 {0x3014, 0x014F},
 {0x321E, 0x0752},
@@ -395,9 +402,11 @@ struct ar0233_reg {
 {0x31BC, 0x0805},
 #if 0
 /* Enable trigger input */
-{0x340A, 0x00E0},
-{0x340C, 0x0002},
-{0x30CE, 0x0120},
+{0x340A, 0x00E0}, // GPIO_CONTROL1: GPIO1 is trigger
+{0x340C, 0x0002}, // GPIO_CONTROL2: GPIO1 is trigger
+{0x30CE, 0x0120}, // TRIGGER_MODE
+//{0x30DC, 0x0120}, // TRIGGER_DELAY
 #endif
+{0x3366, 0x0aaa}, // ANALOG_GAIN
 {0x301A, 0x011C},
 };
diff --git a/drivers/media/i2c/soc_camera/gw4200_ar014x.c b/drivers/media/i2c/soc_camera/gw4200_ar014x.c
index 3bf9d9d..4cda5e5 100644
--- a/drivers/media/i2c/soc_camera/gw4200_ar014x.c
+++ b/drivers/media/i2c/soc_camera/gw4200_ar014x.c
@@ -68,6 +68,7 @@ static void gw4200_s_port(struct i2c_client *client, int fwd_en)
 	};
 }
 
+#if 0
 static int gw4200_set_regs(struct i2c_client *client,
 			  const struct gw4200_reg *regs, int nr_regs)
 {
@@ -84,6 +85,7 @@ static int gw4200_set_regs(struct i2c_client *client,
 
 	return 0;
 }
+#endif
 
 static u16 gw4200_ar014x_read(struct i2c_client *client, u16 addr)
 {
-- 
1.9.1

